### DISCUSSION TITLE
Little return type confusion in LoadClassMetadataEventArgs

### DISCUSSION THREAD

**Discussion by mayerni** [doctrine/doctrine2#11307] [Support Questions]
Hi,

i'm on Symfony 6.4 with Doctrine ORM Bundle. I hope I have the right place to discuss my problem.

I'm currently creating an event listener and hook up to doctrine.event_listener.loadClassMetadata with accessing class Doctrine\ORM\Event\LoadClassMetadataEventArgs. Simplified i make this: 

```php
namespace App\EventListener;

use Doctrine\ORM\Event\LoadClassMetadataEventArgs;

class MyOrmEventListener
{
    [...]
    /**
     * @param \Doctrine\ORM\Event\LoadClassMetadataEventArgs $eventArgs
     *
     * @return void
     */
    public function loadClassMetadata(LoadClassMetadataEventArgs $eventArgs): void
    {
        $classMetadata = $eventArgs->getClassMetadata();
        [...]
    }
}
```

so far so unspectacular... but i have a problem with the return value from getClassMetadata().

In Doctrine\ORM\Event\LoadClassMetadataEventArgs the class said it will use/return **Doctrine\ORM\Mapping\ClassMetadata**:

```php
[...]
namespace Doctrine\ORM\Event;

use Doctrine\ORM\EntityManagerInterface;
use Doctrine\ORM\Mapping\ClassMetadata;
use Doctrine\Persistence\Event\LoadClassMetadataEventArgs as BaseLoadClassMetadataEventArgs;

/**
 * Class that holds event arguments for a loadMetadata event.
 *
 * @extends BaseLoadClassMetadataEventArgs<ClassMetadata<object>, EntityManagerInterface>
 */
class LoadClassMetadataEventArgs extends BaseLoadClassMetadataEventArgs
{
[...]
``` 

But the class it extends from - which is Doctrine\Persistence\Event\LoadClassMetadataEventArgs - it will use/return **Doctrine\Persistence\Mapping\ClassMetadata**, which is not was i looking for:

```php
[...]
namespace Doctrine\Persistence\Event;

use Doctrine\Common\EventArgs;
use Doctrine\Persistence\Mapping\ClassMetadata;
use Doctrine\Persistence\ObjectManager;

/**
 * Class that holds event arguments for a loadMetadata event.
 *
 * @template-covariant TClassMetadata of ClassMetadata<object>
 * @template-covariant TObjectManager of ObjectManager
 */
class LoadClassMetadataEventArgs extends EventArgs
{
    /**
     * @var ClassMetadata
     * @psalm-var TClassMetadata
     */
    private $classMetadata;

    [...]

    public function __construct(ClassMetadata $classMetadata, ObjectManager $objectManager)
    {
        $this->classMetadata = $classMetadata;
        $this->objectManager = $objectManager;
    }
[...]
``` 

Now I'm a bit confused. How do I get the expected return I need? These classes are not 
(fully) interchangeable!? What am I missing? What am I doing wrong?

Thanks for your help.

**Comment by mbabker** [comment]
You should have no issues with this.

Yes, the base event class in the `doctrine/persistence` package expects the base `ClassMetadata` object.  But, the event class also defines generics which will tell anything reading those that if you pass the ORM's event subclass that the event's `getClassMetadata()` method returns the ORM's `ClassMetadata` (similar for the `getObjectManager()` method and specifying whether the ORM's `EntityManagerInterface` or the generic `ObjectManager` is provided).

If something is still groaning about it, you can just add an inline `assert($classMetadata instanceof ORMClassMetadata)` call or an inline docblock for it (I know sometimes PhpStorm doesn't always give the right type info from generics, not sure how well other IDEs support them, so a lot of my code still has those docblocks even though they aren't really needed anymore for any other tooling).

**Reply by mayerni** [reply]
Hello,

many thanks for the answer. The problem I had wasn't that my IDE was giving me something wrong, but rather that I couldn't access the methods "getAssociationMappings()" and "isInheritanceTypeSingleTable" or, more precisely, that PHP throw me a "method not found" error on make:migration.

But now, I got it to work without consciously changing anything, I think.

Sometimes you should just call it a day.

I apologize for the inconvenience.

Thanks alot. :) I will close this.
